### PR TITLE
Feature: CSRF Middleware

### LINKED ISSUE #420: CSRF Middleware | Double Cookie Pattern | Stateless
# Description
Cross-site request forgeries are a type of malicious exploit whereby unauthorized commands are performed on behalf of an authenticated user. We want to be able to provide middleware to prevent such attacks. Since most applications written using zio-http would be stateless we want to be able to provide at least the Double Submit Cookie pattern.

# Double Cookie Pattern
In this technique, we send a random value in both a cookie and as a request parameter, with the server verifying if the cookie value and request value match. When a user visits (even before authenticating to prevent login CSRF), the site should generate a (cryptographically strong) pseudorandom value and set it as a cookie on the user's machine separate from the session identifier. The site then requires that every transaction request include this pseudorandom value as a hidden form value (or other request parameter/header). If both of them match on the server-side, the server accepts it as a legitimate request and if they don't, it would reject the request.

# Api Suggestion
```scala
app @@ csrfBuilder @@ cookie
```

# Specs
* Double Cookie Pattern should be implemented as a middleware
* Unit test cases should be at least 80% for the middleware
* No Scalastyle warnings should be there
* A working example of how to use this middleware

### PR DESCRIPTION
**Pull request by smehta91**
fixes #420

### CODE REVIEW

File: example/src/main/scala/example/CSRF.scala [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import zio.random.Random
+
+object CSRF extends App {
+  val privateApp: HttpApp[Any, Nothing] = Http.collect[Request] { case Method.GET -> !! / "unsafeEndpoint" =>
+    Response.text("secure info")
+  } @@ csrfValidate() // Check for matching csrf header and cookie
+
+  val publicApp: HttpApp[Any, Nothing] = Http.collect[Request] { case Method.GET -> !! / "safeEndpoint" =>
+    Response.text("hello")
+  } @@ csrfGenerate(random.nextString(5).provideLayer(Random.live)) // set x-csrf token cookie
```
Comment (tusharmath) [inline example/src/main/scala/example/CSRF.scala]:
```suggestion
  } @@ csrfGenerate() // set x-csrf token cookie
```

Comment (tusharmath) [reply example/src/main/scala/example/CSRF.scala]: We din't need to provide the env over here.

File: zio-http/src/main/scala/zhttp/http/Middleware.scala [hunk 2]
```diff
@@ -113,6 +113,30 @@ object Middleware {
   def basicAuth(u: String, p: String): Middleware[Any, Nothing] =
     basicAuth { case (user, password) => (user == u) && (password == p) }
 
+  def addCookieM[R, E](cookie: ZIO[R, E, Cookie]): Middleware[R, E] =
+    patchM((_, _) => cookie.mapBoth(Option(_), c => Patch.addHeader(Headers.setCookie(c))))
+
+  def csrfGenerate[R, E](tokenGen: ZIO[R, E, String], tokenName: String = "x-csrf-token"): Middleware[R, E] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Middleware.scala]:
```suggestion
  def csrfGenerate[R, E](tokenGen: ZIO[R, E, String] = random.nextString(6), tokenName: String = "x-csrf-token"): Middleware[R, E] =
```

File: zio-http/src/main/scala/zhttp/http/Middleware.scala [hunk 3]
```diff
@@ -113,6 +113,38 @@ object Middleware {
...
+   * @method
+   *   csrfGenerate - Sets cookie with CSRF token
+   * @method
+   *   csrfValidate - Validate token value in request headers against value in cookies
+   * @see
+   *   https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie
+   */
+
+  def csrfGenerate[R, E](
+    tokenGen: ZIO[R, Nothing, String] = random.nextUUID.map(_.toString),
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Middleware.scala]:
```suggestion
    tokenGen: ZIO[R, Nothing, String] = UIO(java.util.UUID.randomUUID.toString),
```

File: zio-http/src/main/scala/zhttp/http/Middleware.scala [hunk 4]
```diff
@@ -113,6 +113,38 @@ object Middleware {
...
+    tokenName: String = "x-csrf-token",
+  ): Middleware[R, E] =
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/Middleware.scala:136]: I  would also flip the arguments, ie have `token` before the generator.

File: zio-http/src/main/scala/zhttp/http/Middleware.scala [hunk 5]
```diff
@@ -113,6 +113,38 @@ object Middleware {
   def basicAuth(u: String, p: String): Middleware[Any, Nothing] =
     basicAuth { case (user, password) => (user == u) && (password == p) }
 
+  def addCookieM[R, E](cookie: ZIO[R, E, Cookie]): Middleware[R, E] =
+    patchM((_, _) => cookie.mapBoth(Option(_), c => Patch.addHeader(Headers.setCookie(c))))
```
Comment (amitksingh1490) [inline zio-http/src/main/scala/zhttp/http/Middleware.scala]:
```suggestion
    patchZIO((_, _) => cookie.mapBoth(Option(_), c => Patch.addHeader(Headers.setCookie(c))))
```

File: zio-http/src/main/scala/zhttp/http/Middleware.scala (hunk 3 above)
Comment (smehta91) [reply zio-http/src/main/scala/zhttp/http/Middleware.scala]: We won't be able to test this if we make this change.

**Review by amitksingh1490 — APPROVED** [review]

**Pull request merged**